docs/csharp/pattern-matching.md

```diff
@@ -2,31 +2,192 @@
 title: Pattern Matching | C# Guide
 description: Learn about pattern matching expressions in C#
 keywords: .NET, .NET Core, C#
-ms.date: 10/04/2016
+ms.date: 12/09/2016
+ms.author: wiwagn
 ms.topic: article
 ms.prod: .net
 ms.technology: devlang-csharp
 ms.devlang: csharp
 ms.assetid: 1e575c32-2e2b-4425-9dca-7d118f3ed15b
 ---
 
-# 🔧 Pattern Matching
+# Pattern Matching #
 
-> **Note**
-> 
-> This topic hasn’t been written yet! 
->
-> We welcome your input to help shape the scope and approach. You can track the status and provide input on this
-> [issue](https://github.com/dotnet/docs/issues/1114) at GitHub.
-> 
-> If you would like to review early drafts and outlines of this topic, please leave a note with your contact information in the issue.
->
-> Learn more about how you can contribute on [GitHub](https://github.com/dotnet/docs/blob/master/CONTRIBUTING.md).
->
+Pattern matching provides a control flow mechanism that is much more
+generalized than is possible with traditional inheritance hierarchies
+available in Object Oriented practices.
 
-<!--
+You are probably familiar with building class hierarchies and creating
+virtual methods and overriden methods to customize object behavior based
+on the runtime type of the object.
 
-<< Build the sample in an additive way >>
-<< Do the shape thing for the https://docs.microsoft.com/en-us/dotnet/articles/fsharp/language-reference/discriminated-unions shapes >>
+Those techniques aren't possible for data that isn't structured in a class
+hierarchy. In those cases, you'd need to utilize traditional procedural
+imperative constructs like `if` or `switch` statements to manage the control
+flow.
 
--->
\ No newline at end of file
+The new *pattern matching* expressions enable cleaner syntax to examine data
+and manipulate control flow based on any conditon of that data.
+
+In this topic, you'll build a sample class that computes the area of
+different geometric shapes. But, you'll do it without resorting to object
+oriented techniques and building a class hierarchy for the different shapes.
+You'll use *pattern matching* instead. To further emphasize that we're not
+using inheritance, you'll make each shape a `struct` instead of a class. 
```

Comment:
The problem I see with that choice is that this exact example is commonly used to show the benefits of OOP: just inherit from a common base abstract shape type, or implement an IShape interface that has an Area property. Specialize the implementation for each shape, and in the end you get code that is clean and elegant. If you're going to keep that example, you should probably explain in more details the benefits of pattern matching when compared to classic OOP.
One way to do this could be to explain that an application may want to factor itself differently, and separate the concern of computing areas from the concern of storing information about shape dimensions. One example I like more for using pattern matching is serialization: nowadays, people prefer to keep serialization logic separate from the serializable types for obvious reasons: there are many ways, present and future, to serialize objects, and you don't want to put additional burden on type authors because of a concern that is not theirs.
Maybe you can keep the shape example while making it clear that it's a toy model, and maybe pointing to other examples where this makes more sense.